### PR TITLE
Hotfix: TlaOperDecl.isRecursive propagation

### LINKED ISSUE #635: [BUG] Hotfix: TlaOperDecl.isRecursive propagation
Despite the copy-constructor introduced in #347, there are several instances where `TlaOperDecl` are constructed as derivatives  of existing declarations, but without using the copy constructor. Copying this way does not reference the original declaration's `.isRecursive` field, resulting in the flag defaulting to false. For instance, in `IncrementalRenaming`, any operator that is both a) recursive and b) defined inside a LET expression becomes flagged as nonrecursive, triggering a stack overflow in unrolling. 
All such cases need to be found and replaced with copy-constructors. This also future-proofs for #345.

### PR DESCRIPTION
**Pull request by Kukovec**
Closes  #635

- [ ] ~Tests added for any new code~
- [x] Ran `make fmt-fix` (or had formatting run automatically on all files edited)
- [ ] ~Documentation added for any new functionality~
- [ ] ~Entry added to [UNRELEASED.md](./UNRELEASED.md) for any new functionality~

### CODE REVIEW

File: tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/Cacher.scala [hunk 1]
```diff
@@ -5,6 +5,8 @@ import at.forsyte.apalache.tla.lir._
 import at.forsyte.apalache.tla.lir.transformations.{TlaExTransformation, TlaModuleTransformation, TransformationTracker}
 import at.forsyte.apalache.tla.lir.UntypedPredefs._
 
+// TODO: Remove?
+
```
Comment (konnov) [inline tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/Cacher.scala:9]: what is the role of this `Cacher`? Do we still need it? If not, let's remove it.

File: tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/Unroller.scala [hunk 2]
```diff
@@ -157,31 +158,64 @@ class Unroller(nameGenerator: UniqueNameGenerator, tracker: TransformationTracke
...
+  /**
+   * Finds the names of all recursive operators defined under decl.
+   * This includes recursiveoperators defined in LET-IN subexpressions,
+   * as well as the name of `decl` itself, if it is recursive.
+   */
+  private def allRecursiveOperatorsInDecl(decl: TlaOperDecl): Set[String] = {
+    // If this declaration is recursive, we take its base name (for LET-IN defined)
+    val selfOpt = if (decl.isRecursive) Some(IncrementalRenaming.getBase(decl.name)) else None
+    allRecursiveLetInOperatorNames(decl.body) ++ selfOpt
+  }
```
Comment (konnov) [inline tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/Unroller.scala:205]: Does it mean that we do not rely on SANY on identifying recursive operators?

**Review by konnov — APPROVED** [review]
Looks great! I've left a few comments

File: tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/Unroller.scala (hunk 2 above)
Comment (Kukovec) [reply tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/Unroller.scala:205]: We do, but LET-IN defined recursive operators are not toplevel, so you need to scan the entire syntax tree for them.

File: tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/Cacher.scala (hunk 1 above)
Comment (Kukovec) [reply tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/Cacher.scala:9]: I think it was something we disabled a long time ago, I just found it while scanning for instances of OperDecl ctors. If we do remove it, however, that should be its own PR.

**Pull request merged**